kernel/byterun/coq_interp.c

```diff
@@ -422,6 +429,17 @@ value coq_interprete
       }
       Next;
 
+      Instruct(ENSURESTACKCAPACITY) {
+        print_instr("ENSURESTACKCAPACITY");
+        int size = *pc++;
+        if (sp - size < coq_stack_low) {
+          coq_sp = sp;
+          realloc_coq_stack(size * sizeof(value));
+          sp = coq_sp;
+        }
```

Comment:
No, stack doubling is irrelevant here (though it does make the issue less likely to occur). Imagine your stack has size N and it is almost completely filled. Imagine the current function needs about N stack slots before calling another function (and as I explained in the bug-report, this other function does not even need to have arity N for that). So `ENSURESTACKCAPACITY` causes the stack to be reallocated to `2*N` slots. Then the current function tries to perform the call and it reallocates the stack to `4*N` slots because the threshold has been reached. So we are paying two reallocations in a row while we could have reasonably guessed the second reallocation was mandatory because we were about to reach the threshold. The only case where taking the threshold into account in `ENSURESTACKCAPACITY` is a pessimization is when the function does not perform a call (while it required a huge stack).